<div style="width: 100%; overflow: hidden;">
    <a href="http://www.uc.pt/fctuc/dei/">
    <div style="display: block;margin-left: auto;margin-right: auto; width: 100%;"><img src="https://eden.dei.uc.pt/~naml/images_ecos/dei25.png"  /></div>
    </a>
</div>

In [17]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
import seaborn as sns
import torch

%matplotlib inline
# Just to make plots look better
plt.rcParams["figure.figsize"] = (20,12)
plt.rcParams['axes.grid'] = True
plt.style.use('fivethirtyeight')
plt.rcParams['figure.facecolor'] = 'white'
plt.rcParams['axes.facecolor'] = 'white'
plt.rcParams['lines.linewidth'] = 3

<h2><font color='#3498db'>1. Introduction</font></h2>
The main goal of this notebook is introduce the ***PyTorch*** library. PyTorch is an open-source machine learning library primarily developed by Facebook's AI Research lab (FAIR). It provides a flexible and dynamic computational graph, making it popular among researchers and developers for building and training deep learning models. PyTorch supports dynamic computation, which allows you to define and modify the computational graph on-the-fly during runtime. This feature is particularly useful for tasks such as natural language processing and computer vision.
Key features of PyTorch include:

- **Dynamic Computational Graph**: it uses dynamic computation graphs, allowing for easy and intuitive debugging as well as dynamic model architectures.

- **Tensors**: it provides a multi-dimensional array called a tensor, which is similar to NumPy arrays but with the added benefit of GPU acceleration for numerical computations.

- **Autograd**: it includes an automatic differentiation library called Autograd, which automatically computes gradients for tensors during backpropagation. This simplifies the process of training neural networks.

- **Neural Network Module**: it includes the torch.nn module, providing pre-defined layers, loss functions, and optimization algorithms to facilitate the construction of neural network architectures.

- **Eager Execution**: it follows an eager execution model, allowing developers to execute operations immediately as they are defined. This contrasts with other frameworks, like TensorFlow 1.x, which used a static graph.


PyTorch has become a popular choice for researchers and practitioners due to its ease of use, dynamic nature, and strong support for research experimentation. It is widely employed in academia and industry for building and training deep learning models.

*Further Reading:*  
[PyTorch](https://pytorch.org)  



<h2><font color='#3498db'>PyTorch</font></h2>

Let's start by the fundamental concepts of PyTorch. It is recommended that you have a prior understanding of the numpy package since many machine learning frameworks share similar foundational concepts. If you have not encountered numpy before, please refere to the notebook that introduces its functioning.
Lets start by checking the PyTorch version you have available:

In [18]:
torch.__version__

'2.6.0+cpu'

Like many ML frameworks, PyTorch offers stochastic functions, including those for generating random numbers. As such it is recommended that you configure your code for reproducibility by using the same set of random numbers. This is the reason for setting a seed in the code below.

In [19]:
torch.manual_seed(42)

#### PyTorch Tensors
Tensors in PyTorch are the equivalent to Numpy arrays, with the added capability of supporting GPU acceleration (details to be discussed later). 
The term *tensor* is a familiar concept. For example, a vector corresponds to a 1-D tensor, and a matrix to a 2-D tensor. 
In the context of neural networks, tensors of diverse shapes and dimensions become integral to our work.

The most well known functions from numpy can be used on tensors as well. In fact, and given that numpy arrays are so similar to tensors, it is possible to convert them back and forth, but we will not need it too often.

##### Initialization
There are multiple ways to create a tensor. The simplest one is to call `torch.Tensor` with the arguments being the shape of the tensor. Imagine that you would like to create a 2 by 2 tensor. You would do:

In [20]:
x = torch.Tensor(2,2)
x

tensor([[2.3694e-38, 0.0000e+00],
        [0.0000e+00, 0.0000e+00]])

The function torch.Tensor allocates memory for the desired tensor, but reuses any values that have already been in the memory, i.e., the positions are initialized with garbage. To assign values to the tensor during initialization, you could do:
- `torch.zeros`: Creates a tensor filled with zeros
- `torch.ones`: Creates a tensor filled with ones
- `torch.rand`: Creates a tensor with random values uniformly sampled between 0 and 1
- `torch.randn`: Creates a tensor with random values sampled from a normal distribution with mean 0 and variance 1
- `torch.arange`: Creates a tensor containing the values N, N+1, N+2,...,M
- `torch.Tensor` (input list): Creates a tensor from the list elements you provide

In [21]:
# Create a tensor from a (nested) list
x = torch.Tensor([[1, 2], [3, 4]])
x

tensor([[1., 2.],
        [3., 4.]])

In [25]:
# Create a tensor with random values between 0 and 1 with the shape 2 by 2
x = torch.rand(2, 2)
x

tensor([[0.3904, 0.6009],
        [0.2566, 0.7936]])

In [ ]:
# arange: Return evenly spaced values within a given interval. 
# Values are generated within the half-open interval [start, stop). Start is optional
x = torch.arange(15)
x

tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

You can obtain the shape of a tensor in the same way as in numpy (`x.shape`), or using the `.size` method:

In [27]:
shape = x.shape
print("Shape:", x.shape)

size = x.size()
print("Size:", size)

dim1, dim2 = x.size()
print("Size:", dim1, dim2)

Shape: torch.Size([2, 2])
Size: torch.Size([2, 2])
Size: 2 2


In [28]:
x.ndim

2

#### From PyTorch Tensors to Numpy and Back
Tensors can be converted to numpy arrays, and numpy arrays back to tensors. To transform a numpy array into a tensor, we can use the function `torch.from_numpy`:

In [29]:
np_arr = np.array([[1, 2], [3, 4]])
tensor = torch.from_numpy(np_arr)
print("Array:", np_arr)
print("Tensor:", tensor)

Array: [[1 2]
 [3 4]]
Tensor: tensor([[1, 2],
        [3, 4]])


To transform a PyTorch tensor back to a numpy array, we can use the function `.numpy()` on tensors:

In [30]:
tensor = torch.arange(4).view(2,2)
np_arr = tensor.numpy()

print("Tensor:", tensor)
print("Array:", np_arr)

Tensor: tensor([[0, 1],
        [2, 3]])
Array: [[0 1]
 [2 3]]


When converting tensors to NumPy, the tensor needs to be on the CPU rather than the GPU (more details on GPU later). If the tensor is on the GPU, you must first use the `.cpu()` method on the tensor, i.e., the conversion line appears as follows: np_arr = tensor.cpu().numpy().

#### Operations
Most of the operations that exist in numpy exist in [PyTorch](https://pytorch.org/docs/stable/tensors.html#). However but we will see the most important ones bellow, starting with the simplest one: *Addition*

In [31]:
x1 = torch.rand(2, 2)
x2 = torch.rand(2, 2)
add = x1 + x2

print("X1", x1)
print("X2", x2)
print("Result", add)

X1 tensor([[0.9408, 0.1332],
        [0.9346, 0.5936]])
X2 tensor([[0.8694, 0.5677],
        [0.7411, 0.4294]])
Result tensor([[1.8102, 0.7009],
        [1.6757, 1.0230]])


 `x1 + x2` creates a new tensor containing the sum of the two tensors. 
 we can also use in-place operations that are applied directly on the memory of a tensor. For example, we can change the values of `x2` without the chance to re-accessing the values of x2 before the operation.
In-place operations are usually marked with a underscore postfix (e.g. `add_` instead of `add`).

In [32]:
x1 = torch.rand(2, 3)
x2 = torch.rand(2, 3)
print("X1 (before)", x1)
print("X2 (before)", x2)

x2.add_(x1)
print("X1 (after)", x1)
print("X2 (after)", x2)

X1 (before) tensor([[0.8854, 0.5739, 0.2666],
        [0.6274, 0.2696, 0.4414]])
X2 (before) tensor([[0.2969, 0.8317, 0.1053],
        [0.2695, 0.3588, 0.1994]])
X1 (after) tensor([[0.8854, 0.5739, 0.2666],
        [0.6274, 0.2696, 0.4414]])
X2 (after) tensor([[1.1824, 1.4056, 0.3719],
        [0.8969, 0.6284, 0.6407]])


Another frequently used operation involves reshaping a tensor. A tensor with a size of (2, 3) can be rearranged into any other shape with an equivalent number of elements (e.g., tensor of size (6) or (3, 2)). In PyTorch, this operation is referred to as `view`:

In [33]:
x = torch.arange(6)
print("x", x)
x = x.view(2, 3)
print("X", x)
x = x.permute(1, 0) # Swapping dimension 0 and 1
print("X", x)

x tensor([0, 1, 2, 3, 4, 5])
X tensor([[0, 1, 2],
        [3, 4, 5]])
X tensor([[0, 3],
        [1, 4],
        [2, 5]])


Other frequently employed operations involve matrix multiplications, a crucial for neural networks manipulation. Frequently, an input vector `x` is transformed using a learned weight matrix `W`. Various methods and functions for performing matrix multiplication are available:

- `torch.matmul`: Performs the matrix product over two tensors, where the specific behavior depends on the dimensions. If both inputs are matrices (2-dimensional tensors), it performs the standard matrix product. For higher dimensional inputs, the function supports (broadcasting)[https://pytorch.org/docs/stable/generated/torch.matmul.html#torch.matmul]. Can also be written as `a @ b`.
- `torch.mm`: Performs the matrix product over two matrices, but doesn’t support (broadcasting)[https://pytorch.org/docs/stable/generated/torch.mm.html#torch.mm]
- `torch.bmm`: Performs the matrix product with a support batch dimension. If the first tensor *T* is of shape *(b x n x m)*, and the second tensor *R* is (b x m x p), the output *O* is of shape (b x n x p), and has been calculated by performing *b* multipliplications of submatrices of *T* and *R*.

In [34]:
x = torch.arange(6)
x = x.view(2, 3)
print("X", x)

X tensor([[0, 1, 2],
        [3, 4, 5]])


In [35]:
W = torch.arange(9).view(3, 3) # We can also stack multiple operations in a single line
print("W", W)

W tensor([[0, 1, 2],
        [3, 4, 5],
        [6, 7, 8]])


In [36]:
h = torch.matmul(x, W) # Verify the result by calculating it by hand too!
print("h", h)

h tensor([[15, 18, 21],
        [42, 54, 66]])


##### Indexing
Sometimes we need to select a parte of a tensor. Indexing in PyTorch works the same way that in numpy

In [37]:
x = torch.arange(12).view(3, 4)
print("X", x)

X tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])


In [38]:
print(x[:, 1])   # Second column

tensor([1, 5, 9])


In [39]:
print(x[0])      # First row

tensor([0, 1, 2, 3])


In [40]:
print(x[:2, -1]) # First two rows, last column

tensor([3, 7])


In [41]:
print(x[1:3, :]) # Middle two rows

tensor([[ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])


#### Exercise 1.1
Create a 1D tensor of numbers from 0 to 30?


#### Exercise 1.2
Create a 3×3 tensor array with all elements equal to 1.

#### Exercise 1.3
Extract all odd numbers from an tensor.



#### Exercise 1.4
Replace all odd numbers in an tensor with -1.


#### Exercise 1.5
Swap columns 1 and 2 in the tensor x.

x = torch.arange(16).view(4,4)


#### Exercise 1.6
Find the mean, median, standard deviation by row and column of tensor x

x = torch.arange(16).float().view(4,4)